File: stone_processor/src/main/java/com/github/klee0kai/stone/utils/ClassNameUtils.java
package com.github.klee0kai.stone.utils;

import com.github.klee0kai.stone.exceptions.ClassNotFoundStoneException;
import com.github.klee0kai.stone.exceptions.PrimitiveTypeNonSupportedStoneException;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.ParameterizedTypeName;
import com.squareup.javapoet.TypeName;
import com.squareup.javapoet.WildcardTypeName;

import java.util.List;

import static com.github.klee0kai.stone.exceptions.ExceptionStringBuilder.createErrorMes;

/**
 * Class and type resolving utils.
 */
public class ClassNameUtils {

    /**
     * Get class name from full name.
     * Primitives and boxed types are not support
     *
     * @param clFullName class full name like "com.github.klee0kai.stone.annotations.component.class"
     * @return class name
     * @throws PrimitiveTypeNonSupportedStoneException primitive types are not support
     */
    public static ClassName classNameOf(String clFullName) {
        try {
            if (clFullName.endsWith(".class"))
                clFullName = clFullName.substring(0, clFullName.lastIndexOf(".class"));
            return ClassName.get(clFullName.substring(0, clFullName.lastIndexOf(".")),
                    clFullName.substring(clFullName.lastIndexOf(".") + 1));
        } catch (Exception e) {
            if (!clFullName.contains(".")) {
                throw new PrimitiveTypeNonSupportedStoneException(
                        createErrorMes()
                                .primitiveTypesNonSupported(clFullName)
                                .build(),
                        e);
            }
            throw new ClassNotFoundStoneException(
                    createErrorMes()
                            .classNonFound(clFullName)
                            .build(),
                    e);
        }

    }

    public static TypeName rawTypeOf(TypeName typeName) {
        if (typeName instanceof ParameterizedTypeName)
            return rawTypeOf(((ParameterizedTypeName) typeName).rawType);
        if (typeName instanceof WildcardTypeName) {
            List<TypeName> upperBounds = ((WildcardTypeName) typeName).upperBounds;
            return rawTypeOf(!upperBounds.isEmpty() ? upperBounds.get(0) : null);
        }
        return typeName;
    }

    public static String simpleName(TypeName typeName) {
        TypeName raw = rawTypeOf(typeName);
        if (raw instanceof ClassName)
            return ((ClassName) raw).simpleName();
        return null;
    }

}

File: stone_ksp/src/commonMain/kotlin/com/github/klee0kai/thekey/stone/ksp/helpers/wrap/ClassNameUtils.kt
package com.github.klee0kai.thekey.stone.ksp.helpers.wrap

import com.squareup.kotlinpoet.ParameterizedTypeName
import com.squareup.kotlinpoet.TypeName
import com.squareup.kotlinpoet.WildcardTypeName

object ClassNameUtils {

    fun rawTypeOf(typeName: TypeName): TypeName {
        if (typeName is ParameterizedTypeName) {
            return rawTypeOf((typeName).rawType)
        }
        if (typeName is WildcardTypeName) {
            val upperBounds = typeName.outTypes
            if (upperBounds.isNotEmpty()) {
                return rawTypeOf(upperBounds.first())
            }
        }
        return typeName.copy(nullable = false)
    }


    fun noWildCardType(type: TypeName): TypeName {
        if (type is WildcardTypeName) {
            val upperBounds = type.outTypes
            return if (!upperBounds.isEmpty()) {
                noWildCardType(upperBounds.first())
            } else {
                type
            }
        }
        return type
    }

}


fun TypeName.rawType(): TypeName = ClassNameUtils.rawTypeOf(this)